### PR TITLE
Laboratory work #2, Sofya Borodina - 22FPL1

### PR DESCRIPTION
**Pull request by Sofya0borodina**

### CODE REVIEW

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Congratulations with passing the first lab!

There is nothing to review concerning the 2nd lab yet.

**Review by marina-kaz — COMMENTED** [review]
Nothing to review again :<

File: lab_2_tokenize_by_bpe/main.py [hunk 1]
```diff
@@ -36,6 +62,19 @@ def count_tokens_pairs(
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :return: dictionary in the form of <token pair: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict):
+        return None
+    listed_combinations = []
+    combination_frequencies = {}
+    for word in word_frequencies:
+        for i, token in enumerate(word):
+            if (i + 1) < len(word):
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:71]: consider iterating over a range from zero to the desired limit

File: lab_2_tokenize_by_bpe/main.py [hunk 2]
```diff
@@ -36,6 +62,19 @@ def count_tokens_pairs(
...
+                listed_combinations.append((token, word[i + 1]))
+    for combination in listed_combinations:
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:73]: why not add them to the dictionary immediately in the first cycle?

File: lab_2_tokenize_by_bpe/main.py [hunk 3]
```diff
@@ -58,6 +109,40 @@ def train(
...
     """
+    if (not (isinstance(word_frequencies, dict) or word_frequencies is None) or not
+        isinstance(num_merges, int)):
+        return None
+    token_pairs = count_tokens_pairs(word_frequencies)
+    while (num_merges > 0) and len(token_pairs) > 0:
+        num_merges -= 1
+        max_freq = 0
+        max_pairs = []
+        for freq in token_pairs.values():
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:120]: it looks like you are re-implementing the built-in function `max`,  or at least method `sorted`. Consider passing the necessary features via `key` argument.

**Review by marina-kaz — COMMENTED** [review]
Good progress!

File: lab_2_tokenize_by_bpe/main.py [hunk 4]
```diff
@@ -14,6 +14,18 @@ def prepare_word(
     :param end_of_word: a token that signifies the end of word
     :return: preprocessed word
     """
+    if (not isinstance(raw_word, str) or not 
+        (isinstance(start_of_word, str) or start_of_word is None) or not 
+        (isinstance(end_of_word, str) or end_of_word is None)):
+        return None
+    preprocessed_word = []
+    if start_of_word != None:
```
[the remark this answers is not in the document]
Comment (Sofya0borodina) [reply lab_2_tokenize_by_bpe/main.py:22]: Done

File: lab_2_tokenize_by_bpe/main.py [hunk 5]
```diff
@@ -26,6 +38,20 @@ def collect_frequencies(
...
     """
+    if (not isinstance(text, str) or not 
+        (isinstance(start_of_word, str) or start_of_word is None) or not
+        isinstance(end_of_word, str)):
+        return None
+    freq_dictionary = {}
+    listed_text = text.split(" ")
+    for word in listed_text:
+        preprocessed_word = prepare_word(word, start_of_word, end_of_word)
+        if preprocessed_word == None:
```
[the remark this answers is not in the document]
Comment (Sofya0borodina) [reply lab_2_tokenize_by_bpe/main.py:49]: Done

File: lab_2_tokenize_by_bpe/main.py (hunk 1 above)
Comment (Sofya0borodina) [reply lab_2_tokenize_by_bpe/main.py:71]: Fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 2 above)
Comment (Sofya0borodina) [reply lab_2_tokenize_by_bpe/main.py]: Done

File: lab_2_tokenize_by_bpe/main.py [hunk 6]
```diff
@@ -58,6 +106,27 @@ def train(
...
+        isinstance(num_merges, int)):
+        return None
+    merges_count = 0
+    while merges_count < num_merges:
+        token_pairs = count_tokens_pairs(word_frequencies)
+        if token_pairs is None:
+            return None
+        if token_pairs != {}:
+            new_dict = dict(
+                sorted(token_pairs.items(),
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:119]: Sorting a dictionary 👎

File: lab_2_tokenize_by_bpe/main.py [hunk 7]
```diff
@@ -58,6 +106,27 @@ def train(
...
+                       key=lambda item: (-item[1], -len("".join(item[0])), "".join(item[0]))))
+            for k in new_dict.keys():
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:121]: What about `num_merges`? What if we only needed to merge, like, three pairs, and here you are merging everything...

**Review by marina-kaz — COMMENTED** [review]
Good progress, I see the effort!

Keep going!

File: lab_2_tokenize_by_bpe/main.py (hunk 6 above)
Comment (Sofya0borodina) [reply lab_2_tokenize_by_bpe/main.py]: I'm sorry, I didn't find the solution without sorting

File: lab_2_tokenize_by_bpe/main.py [hunk 8]
```diff
@@ -58,6 +106,27 @@ def train(
...
+                word_frequencies = merge_tokens(word_frequencies, k)
+                if word_frequencies is None:
+                    return None
+                merges_count += 1
+                break
+            continue
+        break
```
[the remark this answers is not in the document]
Comment (Sofya0borodina) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 7 above)
Comment (Sofya0borodina) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py [hunk 9]
```diff
@@ -58,6 +106,27 @@ def train(
...
     """
+    if (not (isinstance(word_frequencies, dict) or word_frequencies is None) or not
+        isinstance(num_merges, int)):
+        return None
+    merges_count = 0
+    while merges_count < num_merges:
+        token_pairs = count_tokens_pairs(word_frequencies)
+        if token_pairs is None:
+            return None
+        if token_pairs != {}:
```
[the remark this answers is not in the document]
Comment (Sofya0borodina) [reply lab_2_tokenize_by_bpe/main.py]: fixed 1

File: lab_2_tokenize_by_bpe/main.py [hunk 10]
```diff
@@ -47,6 +82,22 @@ def merge_tokens(
...
+        isinstance(pair, tuple)):
+        return None
+    merged_frequencies = {}
+    for word, freq in word_frequencies.items():
+        if ''.join(pair) in ''.join(word):
+            listed_word = list(word)
+            for i in range(len(listed_word) - 1):
+                if (listed_word[i], listed_word[i + 1]) == pair:
+                    listed_word[i + 1] = pair[0] + pair[1]
+                    listed_word[i] = ''
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:95]: Just pop an element by its index

File: lab_2_tokenize_by_bpe/main.py [hunk 11]
```diff
@@ -69,6 +137,20 @@ def get_vocabulary(
...
+        return None
+    listed_tokens = set()
+    token_identifier_dict = {}
+    for word in word_frequencies:
+        for token in word:
+            listed_tokens.add(token)
+            for element in token:
+                listed_tokens.add(element)
+    listed_tokens.add(unknown_token)
+    sorted_tokens = sorted(listed_tokens, key=lambda x: (-len(x), x))
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:150]: Set are unordered structures, let's not sort it

File: lab_2_tokenize_by_bpe/main.py [hunk 12]
```diff
@@ -81,6 +163,18 @@ def decode(
...
     :return: decoded sequence
     """
+    if not isinstance(encoded_text, list) or not isinstance(vocabulary, dict) or not (isinstance(
+            end_of_word_token, str) or end_of_word_token is None):
+        return None
+    decoded_tokens = []
+    for i in encoded_text:
+        for token, token_i in vocabulary.items():
+            if token_i == i and end_of_word_token is not None:
+                decoded_tokens.append(' ' if token == end_of_word_token else token)
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:173]: I suggest simply replacing the EoW token in the resulting string

**Review by marina-kaz — COMMENTED** [review]
Good job!

File: lab_2_tokenize_by_bpe/main.py (hunk 10 above)
Comment (Sofya0borodina) [reply lab_2_tokenize_by_bpe/main.py]: Done

File: lab_2_tokenize_by_bpe/main.py (hunk 11 above)
Comment (Sofya0borodina) [reply lab_2_tokenize_by_bpe/main.py]: Fixed (If I understood the problem correctly)

File: lab_2_tokenize_by_bpe/main.py (hunk 12 above)
Comment (Sofya0borodina) [reply lab_2_tokenize_by_bpe/main.py:174]: I'm sorry, I didn't quite understand what exactly is required of me.

**Pull request closed without merging**